RIOT-OS/RIOT
# drivers/at: fix URC handling and add better testing

### Contribution description
Changes in this PR:
  - remove asynchronous URC handling, as it is broken (see https://github.com/RIOT-OS/RIOT/issues/20059). I might get back to this feature in the future and re-implement it, but it would need a complete re-work.
  - add robust URC handling where it's possible and document which actions the user has to take where it's not
  - refactor the response parsing to make it individually testable and add extensive unit testing.
 
At the first glance, the diffs in this PR are huge. However, much of it comes from refactoring. Therefore, I have split this PR in three commits:
 1. [drivers/at: removed async URC handling.](https://github.com/RIOT-OS/RIOT/commit/735edd1ec37769ef47a5f68ce5fd58c77698a793) - removes asynchronous URC handling. **I am not sure I properly removed this feature from everywhere, please point me to  additional locations that I might have overlooked.**
 2. [drivers/at: refactor for better unit testing of parsing methods](https://github.com/RIOT-OS/RIOT/commit/9f542c6c9b3f25eff4acd5af229600f07075a0d2) - extracts response parsing from the command sending procedures in the scope of unit testing
 3. [drivers/at: URC handling done, added unit tests, documentation.](https://github.com/RIOT-OS/RIOT/commit/abcb50f2065f90ac8c341634f16b7bb600686048) - this is where the actual functional changes happen. Adds roubst URC handling, unit testing and proper documentation.




### Testing procedure
1. See `tests/drivers/at/tests-with-config/emulate_dce.py`
2. On native, run `tests/drivers/tests-with-config/test_all_configs.sh`. This will compile the `at_tests` program and run the unit tests with all possible configuration options. No actual modem is needed. 



### Issues/PRs references

addresses https://github.com/RIOT-OS/RIOT/issues/20059
removes https://github.com/RIOT-OS/RIOT/pull/14327#issuecomment-651788427
